src/Lean/Compiler/IR/EmitLLVM.lean

```diff
@@ -0,0 +1,1521 @@
+        let _ ← LLVM.buildRet builder retv
+        pure ShouldForwardControlFlow.no
+      else do
+        callLeanDecRef builder resv
+        let _ ← LLVM.buildRet builder (← LLVM.constInt64 llvmctx 0)
+        pure ShouldForwardControlFlow.no
+
+    )
+    (fun builder => do -- else builder
+        let resv ← LLVM.buildLoad2 builder resty res "resv"
+        callLeanIOResultShowError builder resv
+        callLeanDecRef builder resv
+        let _ ← LLVM.buildRet builder (← LLVM.constInt64 llvmctx 1)
+        pure ShouldForwardControlFlow.no)
+  -- at the merge
+  let _ ← LLVM.buildUnreachable builder
+
+def hasMainFn : M llvmctx Bool := do
+  let env ← getEnv
+  let decls := getDecls env
+  return decls.any (fun d => d.name == `main)
+
+def emitMainFnIfNeeded (mod : LLVM.Module llvmctx) (builder : LLVM.Builder llvmctx) : M llvmctx Unit := do
+  if (← hasMainFn) then emitMainFn mod builder
+
+def main : M llvmctx Unit := do
+  emitFnDecls
+  let builder ← LLVM.createBuilderInContext llvmctx
+  emitFns (← getLLVMModule) builder
+  emitInitFn (← getLLVMModule) builder
+  emitMainFnIfNeeded (← getLLVMModule) builder
+end EmitLLVM
+
+def getLeanHBcPath : IO System.FilePath := do
+  return (← getLibDir (← getBuildDir)) / "lean.h.bc"
+
+def optimizeLLVMModule (mod : LLVM.Module ctx) : IO Unit := do
+  let pm  ← LLVM.createPassManager
+  let pmb ← LLVM.createPassManagerBuilder
+  pmb.setOptLevel 3
+  pmb.populateModulePassManager pm
+  LLVM.runPassManager pm mod
+  LLVM.disposePassManager pm
+  LLVM.disposePassManagerBuilder pmb
+
+@[export lean_ir_emit_llvm]
+def emitLLVM (env : Environment) (modName : Name) (filepath : String) : IO Unit := do
+  LLVM.llvmInitialize
+  let llvmctx ← LLVM.createContext
+  let module ← LLVM.createModule llvmctx modName.toString
+  let emitLLVMCtx : EmitLLVM.Context llvmctx := {env := env, modName := modName, llvmmodule := module}
+  let initState := { var2val := default, jp2bb := default : EmitLLVM.State llvmctx}
+  let out? ← (EmitLLVM.main (llvmctx := llvmctx) initState).run emitLLVMCtx
+  match out? with
+  | .ok _ => do
+         let membuf ← LLVM.createMemoryBufferWithContentsOfFile (← getLeanHBcPath).toString
+         let modruntime ← LLVM.parseBitcode llvmctx membuf
+         LLVM.linkModules (dest := emitLLVMCtx.llvmmodule) (src := modruntime)
+         optimizeLLVMModule emitLLVMCtx.llvmmodule
+         LLVM.writeBitcodeToFile emitLLVMCtx.llvmmodule filepath
+         let tripleStr ← LLVM.getDefaultTargetTriple
+         let target ← LLVM.getTargetFromTriple tripleStr
+         let cpu := "generic"
+         let features := ""
+         let targetMachine ← LLVM.createTargetMachine target tripleStr cpu features
+         let codegenType := LLVM.CodegenFileType.ObjectFile
+         LLVM.targetMachineEmitToFile targetMachine emitLLVMCtx.llvmmodule (filepath ++ ".o") codegenType
+         LLVM.disposeModule emitLLVMCtx.llvmmodule
+         LLVM.disposeTargetMachine targetMachine
```

Comment:
Yes, I agree it is unlikely to go wrong, since they are allocated once and freed once. I did make an issue to track this: https://github.com/leanprover/lean4/issues/1974